### PR TITLE
Remove C APIs and "CTAP1" argument variants

### PR DESCRIPTION
**Pull request by jschanck**
Time to remove some code!

Firefox's implementation of the legacy U2F javascript API was the last consumer of authenticator's C API. Now that Firefox has removed the legacy U2F javascript API, we can remove the C API.

We can also remove `RegisterArgs::CTAP1` and `SignArgs::CTAP1`. A WebAuthn caller always has enough information to give us the CTAP2 variants. This would let us replace `StateMachine` by `StateMachineCtap2` if we did not still need `StateMachine` for Firefox users who rely on `security.webauthn.ctap2=false`.

To accomodate the `ctap2=false` users, I've moved the old `StateMachine::register` and `StateMachine::sign` to `StateMachineCtap2::legacy_register` and `StateMachineCtap2::legacy_sign`. I added a flag to `RegisterArgs` and `SignArgs` to signal when requests should be downgraded.  With that change, it's safe to remove the existing `StateMachine` and rename `StateMachineCtap2` to `StateMachine`.

I don't want to handle `RegisterResult::CTAP1` and `SignResult::CTAP1` values from the legacy functions in `authrs_bridge`, so `legacy_register` and `legacy_sign` repackage them. There's a small risk that this breaks some clients who are relying on `security.webauthn.ctap2=false`, but any breakages will be easy to fix.

The CTAP2 interface's use of client data JSON, instead of its hash, got in the way of unifying the result handling, so I took care of that too.

### CODE REVIEW

File: src/ctap2/commands/get_assertion.rs [hunk 1]
```diff
@@ -717,7 +695,9 @@ pub mod test {
             token_binding: Some(TokenBinding::Present(String::from("AAECAw"))),
         };
         let assertion = GetAssertion::new(
-            client_data.clone(),
+            CollectedClientDataWrapper::new(client_data)
```
Comment (msirringhaus) [inline src/ctap2/commands/get_assertion.rs]: Do we still need `CollectedClientDataWrapper`?
I think we can also remove that and simply have `CollectedClientData` with a `hash()`-function.

File: src/lib.rs [hunk 2]
```diff
@@ -92,13 +85,13 @@ pub type AppId = Vec<u8>;
 #[derive(Debug)]
 pub enum RegisterResult {
     CTAP1(Vec<u8>, u2ftypes::U2FDeviceInfo),
-    CTAP2(AttestationObject, CollectedClientDataWrapper),
+    CTAP2(AttestationObject),
```
Comment (msirringhaus) [inline src/lib.rs:88]: Do we still need `RegisterResult` (and `SignResult`) with the `CTAP1` and `CTAP2`-variants? Isn't it now always `CTAP2`, aka `AttestationObject` here and `AssertionObject` for signing?

File: src/authenticatorservice.rs [hunk 3]
```diff
@@ -214,68 +147,7 @@ impl AuthenticatorService {
...
-        args: RegisterArgsCtap1,
-        status: Sender<crate::StatusUpdate>,
-        callback: StateCallback<crate::Result<crate::RegisterResult>>,
-    ) -> crate::Result<()> {
-        if args.challenge.len() != PARAMETER_SIZE || args.application.len() != PARAMETER_SIZE {
-            return Err(AuthenticatorError::InvalidRelyingPartyInput);
-        }
-
-        for key_handle in &args.key_handles {
-            if key_handle.credential.len() >= 256 {
```
Comment (msirringhaus) [inline src/authenticatorservice.rs:239]: About removing those checks:
It seems, that `u2f_is_keyhandle_valid()` is handling these as well. We will not error out, though, but silently ignore those that are not valid (which I think is ok).
**But** `u2f_is_keyhandle_valid()` has an off-by-one error in that check: It checks for `> 256`, when it should be `>= 256`.

Comment (jschanck) [reply src/authenticatorservice.rs:239]: Good catch!

File: src/ctap2/commands/get_assertion.rs (hunk 1 above)
Comment (jschanck) [reply src/ctap2/commands/get_assertion.rs]: Yes, we can remove it.

File: src/lib.rs (hunk 2 above)
Comment (jschanck) [reply src/lib.rs:88]: The `CTAP1` variant is still used in `virtualdevices/software_u2f.rs`. I'm going to update that soon, and I will simplify `RegisterResult` and `SignResult` at the same time.

File: src/authenticatorservice.rs [hunk 4]
```diff
@@ -596,276 +403,39 @@ mod tests {
...
-
-        let large_key = KeyHandle {
-            credential: vec![0; 1000],
-            transports: AuthenticatorTransports::USB,
-        };
-
-        let mut s = AuthenticatorService::new(CtapVersion::CTAP2).unwrap();
-        s.add_transport(Box::new(TestTransportDriver::new(true).unwrap()));
-
-        let ctap2_register_args = RegisterArgsCtap2 {
```
Comment (mozkeeler) [inline src/authenticatorservice.rs:805]: Is this test not still relevant?

**Review by mozkeeler — APPROVED** [review]
LGTM!

File: src/authenticatorservice.rs (hunk 4 above)
Comment (jschanck) [reply src/authenticatorservice.rs:805]: It sort of made sense to test that `AuthenticatorService::register_ctap2` accepted large keys when there was an `AuthenticatorService::register_ctap1` function that rejected large keys. Now there's only one `register` function, and it accepts large keys, so I don't think the test is necessary.

**Pull request merged**